lib/web_ui/lib/src/engine/dom.dart

```diff
@@ -756,56 +766,400 @@ extension DomCanvasGradientExtension on DomCanvasGradient {
 @staticInterop
 class DomXMLHttpRequestEventTarget extends DomEventTarget {}
 
-@JS()
-@staticInterop
-class DomXMLHttpRequest extends DomXMLHttpRequestEventTarget {}
+Future<_DomResponse> _rawHttpGet(String url) {
+  return js_util.promiseToFuture<_DomResponse>(js_util.callMethod(domWindow, 'fetch', <String>[url]));
+}
+
+typedef MockHttpFetchResponseFactory = Future<MockHttpFetchResponse> Function(String url);
+
+MockHttpFetchResponseFactory? mockHttpFetchResponseFactory;
+
+/// Makes an HTTP GET request to the given [url] and returns the response.
+///
+/// If the request fails, throws [HttpFetchError]. HTTP error statuses, such as
+/// 404 and 500 are not treated as request failures. In those cases the HTTP
+/// part did succeed and correctly passed the HTTP status down from the server
+/// to the client. Those statuses represent application-level errors that need
+/// extra interpretation to decide if they are "failures" or not. See
+/// [HttpFetchResponse.hasPayload] and [HttpFetchResponse.payload].
+///
+/// This function is designed to handle the most general cases. If the default
+/// payload handling, including error checking, is sufficient, consider using
+/// convenience functions [httpFetchByteBuffer], [httpFetchJson], or
+/// [httpFetchText] instead.
+Future<HttpFetchResponse> httpFetch(String url) async {
+  if (mockHttpFetchResponseFactory != null) {
+    return mockHttpFetchResponseFactory!(url);
+  }
+  try {
+    final _DomResponse domResponse = await _rawHttpGet(url);
+    return HttpFetchResponseImpl._(url, domResponse);
```

Comment:
Since _DomResponse is a `@staticInterop` object, you may be able to simplify this a bunch by just making the `mockHttpFetchResponseFactory` return a

```
js_util.jsify({
  // Something that resembles a _DomResponse
}) as _DomResponse;
```

(Or even create a real _DomResponse object with its [constructor](https://developer.mozilla.org/en-US/docs/Web/API/Response/Response)) 

That would allow you to simplify the objects returned by fetch quite a bit, and also:

```
try {
  _DomResponse response;
  if (mockHttpFetchResponseFactory != null) {
    response = await mockHttpFetchResponseFactory!(url);
  } else {
    response = await _rawHttpGet(url);
  }
  return HttpFetchResponseImpl._(url, domResponse);
} catch (...) {
  //...
}
```

test what happens when the mockHttpRequestFactory throws.